src/commands/migrate.json

```diff
@@ -56,24 +56,17 @@
                 }
             },
             {
+                "notes": "For the optional KEYS keyword. It is marked 'unknown' because the word 'KEYS' may also appear when using AUTH and AUTH2.",
                 "flags": [
                     "RW",
                     "ACCESS",
-                    "DELETE",
-                    "INCOMPLETE"
```

Comment:
@CharlesChen888 is the change in the json file really necessary? it seems like the bug is only in `migrateGetKeys`

Reply:
We can take a look at what is said in redis.io:

> Thus, we recognize that key specifications are incomplete and may fail to extract all keys. However, we assure that even incomplete specifications never yield the wrong names of keys, providing that the command is syntactically correct.
> 
> In the case of [MIGRATE](https://redis.io/commands/migrate), the search begins at the end (startfrom has the value of -1). If and when we encounter a key named "KEYS", we'll only extract the subset of the key name arguments after it. That's why [MIGRATE](https://redis.io/commands/migrate) has the incomplete flag in its key specification.

"INCOMPLETE" means only that Redis may miss some keys or keywords. So Redis should assure that not any wrong key would be extracted when calling `MIGRATE`. However, before using `migrateGetKeys`, function `getKeysFromCommandWithSpecs` always tries to use the `key_specs` in the json file, and it may get wrong keys. Although it finally uses `migrateGetKeys`(as it is marked "INCOMPLETE") to cover the results, this still could be a hidden trouble.

Then we can take a look at what is said about the `SORT` command:

> Another case of incompleteness is the [SORT](https://redis.io/commands/sort) command. Here, the begin_search and find_keys are of type unknown. The client should revert to calling the [COMMAND GETKEYS](https://redis.io/commands/command-getkeys) command to extract key names from the arguments, short of implementing it natively. The difficulty arises, for example, because the string "STORE" is both a keyword (token) and a valid literal argument for [SORT](https://redis.io/commands/sort).

Similar to `MIGRATE`, but it is marked "UNKNOWN", as it means we are not sure where the keywords are. Therefore, since now `MIGRATE` does not only have incomplete key specification, it may also get wrong keys, I think it is necessary to change the json file.

Reply:
i get your point.

the only reason we chose to give actual key-specs to MIGRATE is that it is possible to detect some of the keys without false negatives, while in SORT it's impossible (because key-specs are limited)

now it got me thinking if there's even a point in giving a method that can only detect some of the keys. how would that be even useful? i can't imagine a use case where only some of the keys are sufficient.

i'm leaning towards taking an "all or nothing" approach: either you can count on key-specs to detect all keys, or use COMMAND GETKEYS (just to put things in proportion: the only problematic commands are MIGRATE and SORT)

if we got that way, we can drop some ad-hoc code we added in order to partially support MIGRATE:
1. completely get rid of the INCOMPLETE key-spec flag (somewhat breaking but i guess it's ok)
2. completely get rid of negative `startfrom`

@itamarhaber @yossigo WDYT?

Reply:
> now it got me thinking if there's even a point in giving a method that can only detect some of the keys. how would that be even useful? i can't imagine a use case where only some of the keys are sufficient.

cluster aware clients need just one key in order to route the command to the right node. and a password or user name called KEYS is probably not a realistic case.

> The "all or nothing" approach can work well (not only for MIGRATE), so I vote for getting rid of the INCOMPLETE flag.

as long as we can avoid false positives, i don't think the all or nothing approach is critical.
i think a client that only needs one key, can run the key-spec algorithm on a command, and only if it yields 0 keys, it resorts to an extra round trip.
for a workload that's heavy of MIGRATE commands it could make a big difference to avoid an extra round trip on every call.

on the other hand, the client that really cares, can implement a specific code to extract the keys for MIGRATE, rather than rely on key-specs. we need to remember that the main motivation for key-specs was to implicitly support new commands that are introduced (we won't have any more MIGRATEs), **but also to implicitly support module commands**.
so maybe module commands are a reason to keep it?

**i think the discussion boils down to whether or not it produces false positives**.
which we meant not to produce, but considering for how long we cooked this change, but some intention got lost in the details.

as far as i can tell, the code in `getKeysUsingKeySpecs` doesn't return any false positives. the problem with some of the tests in this PR is just because of the bug in `migrateGetKeys` and the fallback in getKeysFromCommandWithSpecs to use `migrateGetKeys` due to the INCOMPLETE flag, but that doesn't really imply anything on client libraries.

i personally don't see this as a reason to go back on our decision to add the INCOMPLETE flag. all i see here is a bug in `migrateGetKeys` that is misfortune and also affects `getKeysFromCommandWithSpecs`